### PR TITLE
sync origin authz resources to rbac

### PR DESCRIPTION
**Pull request by deads2k**
Still needs tests and I haven't filled in for clusterrolebindings or rolebindings, but you can see where I'm going.

Also, something seems to be wrong.  I'm seeing this every second:
```
E0504 16:21:18.835116   12660 streamwatcher.go:109] Unable to decode an event from the watch stream: got short buffer with n=0, base=53072, cap=163840
W0504 16:21:18.835149   12660 reflector.go:323] github.com/openshift/origin/pkg/authorization/generated/informers/internalversion/factory.go:45: watch of *api.ClusterRole ended with: very short watch
```

@sttts @ncdc @liggitt

### CODE REVIEW

**Comment by deads2k** [comment]
@enj want to do a controller pull?

**Comment by enj** [comment]
> want to do a controller pull?

I think I am a little over committed right now / need to actually finish the things I already have on my todo list.

**Comment by deads2k** [comment]
I wonder if we're using the generated informers anywhere. Some seems to be up with them

```
W0505 15:25:03.351904   13029 reflector.go:323] github.com/openshift/origin/pkg/authorization/generated/informers/internalversion/factory.go:45: watch of *api.ClusterRoleBinding ended with: very short watch
E0505 15:25:03.488272   13029 streamwatcher.go:109] Unable to decode an event from the watch stream: got short buffer with n=0, base=38193, cap=163840
W0505 15:25:03.488297   13029 reflector.go:323] github.com/openshift/origin/pkg/authorization/generated/informers/internalversion/factory.go:45: watch of *api.ClusterRole ended with: very short watch
E0505 15:25:04.328323   13029 streamwatcher.go:109] Unable to decode an event from the watch stream: got short buffer with n=0, base=491, cap=2560
W0505 15:25:04.328353   13029 reflector.go:323] github.com/openshift/origin/pkg/authorization/generated/informers/internalversion/factory.go:45: watch of *api.Role ended with: very short watch
E0505 15:25:04.328896   13029 streamwatcher.go:109] Unable to decode an event from the watch stream: got short buffer with n=0, base=3332, cap=8448
W0505 15:25:04.328912   13029 reflector.go:323] github.com/openshift/origin/pkg/authorization/generated/informers/internalversion/factory.go:45: watch of *api.RoleBinding ended with: very short watch
E0505 15:25:04.360537   13029 streamwatcher.go:109] Unable to decode an event from the watch stream: got short buffer with n=0, base=14112, cap=40960
```

Problem for monday :)

**Comment by deads2k** [comment]
tests added. ready for review.

**Comment by mfojtik** [comment]
@soltysh can you help reviewing this?

File: pkg/cmd/server/origin/master_config.go [hunk 1]
```diff
@@ -223,6 +231,7 @@ func BuildMasterConfig(options configapi.MasterConfig) (*MasterConfig, error) {
 	const defaultInformerResyncPeriod = 10 * time.Minute
 	internalkubeInformerFactory := kinternalinformers.NewSharedInformerFactory(privilegedLoopbackKubeClientsetInternal, defaultInformerResyncPeriod)
 	externalkubeInformerFactory := kinformers.NewSharedInformerFactory(privilegedLoopbackKubeClientsetExternal, defaultInformerResyncPeriod)
+	authorizationInformers := authorizationinformer.NewSharedInformerFactory(authorizationClient, 5*time.Minute)
```
Comment (soltysh) [inline pkg/cmd/server/origin/master_config.go]: Can you please add a TODO here, we should have a single place where we'll start all the informers (like k8s has), I've already opened https://github.com/openshift/origin/issues/14108 to address the switch.

File: pkg/authorization/controller/authorizationsync/origin_to_rbac_test.go [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+			startingOrigin: []*authorizationapi.ClusterRole{
+				{ObjectMeta: metav1.ObjectMeta{Name: "resource-01"}},
+			},
+			actionCheck: func(actions []clienttesting.Action) error {
+				if len(actions) != 0 {
+					return fmt.Errorf("expected %v, got %v", 0, actions)
+				}
+				return nil
+			},
+		},
```
Comment (soltysh) [inline pkg/authorization/controller/authorizationsync/origin_to_rbac_test.go]: I'm missing following tests:
- for `openshift.io/reconcile-protect` annotation
- simple update, but in the other direction, iow. rbac role has changed, but should be overriden
- the invalid update

File: pkg/authorization/controller/authorizationsync/origin_to_rbac_test.go [hunk 3]
```diff
@@ -0,0 +1,545 @@
+package authorizationsync
```
Comment (soltysh) [inline pkg/authorization/controller/authorizationsync/origin_to_rbac_test.go]: \<nit pick mode on\>
It's convenient to have a controller and accompanying test along with it, can you split this into 4 files, please 🙏 
\</nit pick mode off\>

File: pkg/authorization/controller/authorizationsync/origin_to_rbac_test.go (hunk 2 above)
Comment (soltysh) [inline pkg/authorization/controller/authorizationsync/origin_to_rbac_test.go]: This applies to other controllers as well (except for the annotation, which is only for `ClusterRoles`).

**Comment by soltysh** [comment]
@deads2k are you thinking about having integration or extended tests verifying this all in action?

File: pkg/authorization/controller/authorizationsync/generic.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kubernetes/pkg/controller"
+)
+
+type genericController struct {
```
Comment (sttts) [inline pkg/authorization/controller/authorizationsync/generic.go]: godoc

File: pkg/authorization/controller/authorizationsync/origin_to_rbac_clusterrolebinding_controller.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+func NewOriginToRBACClusterRoleBindingController(rbacClusterRoleBindingInformer rbacinformers.ClusterRoleBindingInformer, originClusterRoleBindingInformer origininformers.ClusterRoleBindingInformer, rbacClient rbacclient.ClusterRoleBindingsGetter) *OriginClusterRoleBindingToRBACClusterRoleBindingController {
+	c := &OriginClusterRoleBindingToRBACClusterRoleBindingController{
+		rbacClient:   rbacClient,
+		rbacLister:   rbacClusterRoleBindingInformer.Lister(),
+		originLister: originClusterRoleBindingInformer.Lister(),
+
+		genericController: genericController{
+			name: "OriginClusterRoleBindingToRBACClusterRoleBindingController",
```
Comment (sttts) [inline pkg/authorization/controller/authorizationsync/origin_to_rbac_clusterrolebinding_controller.go]: What a name :)

**Comment by sttts** [comment]
fyi, disabling the rbac bootstrap hook works nearly fine in https://github.com/openshift/origin/pull/14002 with this PR, at least the escalation issue is gone. The new controllers though remove new rbac objects immediately and the etcd storage path tests complain about failed cleanup (of course the etcd storage test itself races with the controllers as well).

@enj, any idea what we do here? We could just create the test objects in the origin auth group and wait until the they show up as rbac rules correctly in the storage?

**Comment by deads2k** [comment]
> @deads2k are you thinking about having integration or extended tests verifying this all in action?

I wasn't planning on it.  I would expect an `oadm migrate` utility to check to make sure everything has been cloned successfully, so perhaps an integration test of that would do.  We will count on these being up to date when we switch over to RBAC authorization.

File: pkg/authorization/controller/authorizationsync/origin_to_rbac_clusterrolebinding_controller.go (hunk 5 above)
Comment (deads2k) [inline pkg/authorization/controller/authorizationsync/origin_to_rbac_clusterrolebinding_controller.go]: > What a name :)

You're just jealous that I get the best ones.

File: pkg/cmd/server/origin/master_config.go (hunk 1 above)
Comment (deads2k) [inline pkg/cmd/server/origin/master_config.go]: > why not defaultInformerResyncPeriod?

fixed.

**Comment by deads2k** [comment]
comments addressed, test added.

**Comment by deads2k** [comment]
[test]

**Comment by soltysh** [comment]
LGTM

**Comment by deads2k** [comment]
This is blocking a pull which fixes a bug [merge]

**Comment by deads2k** [comment]
re[test]

File: test/integration/etcd_storage_path_test.go [hunk 6]
```diff
@@ -842,14 +841,18 @@ func testEtcdStoragePath(t *testing.T, etcdServer *etcdtest.EtcdTestServer, gett
 	if err != nil {
 		t.Fatalf("error getting master config: %#v", err)
 	}
-	masterConfig.AdmissionConfig.PluginOrderOverride = []string{"PodNodeSelector"} // remove most admission checks to make testing easier
+	masterConfig.AdmissionConfig.PluginOrderOverride = []string{"PodNodeSelector"}                        // remove most admission checks to make testing easier
+	masterConfig.KubernetesMasterConfig.AdmissionConfig.PluginOrderOverride = []string{"PodNodeSelector"} // remove most admission checks to make testing easier
+	masterConfig.AdmissionConfig.PluginConfig["ServiceAccount"] = serverapi.AdmissionPluginConfig{
```
Comment (deads2k) [inline test/integration/etcd_storage_path_test.go]: @enj bet you were wondering why you needed to wait for the service accounts...

**Comment by deads2k** [comment]
ansible set up re[test]

**Comment by soltysh** [comment]
Looks like another instance of https://github.com/openshift/origin/issues/14161

**Comment by smarterclayton** [comment]
Not positive it was this change, but after it merged we went from 1hr for test check to 1h53m

**Comment by smarterclayton** [comment]
Yeah, it was this change.  test-cmd got about 25-40% slower - here's a random test, but all of them show it: https://ci.openshift.redhat.com/jenkins/job/test_branch_origin/492/testReport/github/com_openshift_origin_test_cmd_deployments_setdeploymenthook/history/

File: pkg/cmd/server/origin/run_components.go [hunk 7]
```diff
@@ -595,3 +596,31 @@ func (c *MasterConfig) RunUnidlingController() {
 func (c *MasterConfig) RunTemplateController() {
 	go templatecontroller.NewTemplateInstanceController(c.PrivilegedLoopbackClientConfig).Run(utilwait.NeverStop)
 }
+
+func (c *MasterConfig) RunOriginToRBACSyncControllers() {
+	clusterRoles := authorizationsync.NewOriginToRBACClusterRoleController(
```
Comment (smarterclayton) [inline pkg/cmd/server/origin/run_components.go]: How many new informers is this?

**Comment by smarterclayton** [comment]
Looks like this is looping trying to WATCH things that don't support watches:

```
ernetes/5115d70",code="200",contentType="application/vnd.kubernetes.protobuf",resource="clusterrolebindings",verb="WATCH"} 329
ernetes/5115d70",code="200",contentType="application/vnd.kubernetes.protobuf",resource="clusterroles",verb="DELETE"} 30
ernetes/5115d70",code="200",contentType="application/vnd.kubernetes.protobuf",resource="clusterroles",verb="GET"} 39
ernetes/5115d70",code="200",contentType="application/vnd.kubernetes.protobuf",resource="clusterroles",verb="LIST"} 3
ernetes/5115d70",code="200",contentType="application/vnd.kubernetes.protobuf",resource="clusterroles",verb="PUT"} 9
ernetes/5115d70",code="200",contentType="application/vnd.kubernetes.protobuf",resource="clusterroles",verb="WATCH"} 329
```

Which shows up as:

```
E0514 21:35:26.248756   71582 streamwatcher.go:109] Unable to decode an event from the watch stream: got short buffer with n=0, base=13830, cap=40960
W0514 21:35:26.248805   71582 reflector.go:323] github.com/openshift/origin/pkg/authorization/generated/informers/internalversion/factory.go:45: watch of *api.RoleBinding ended with: very short watch
E0514 21:35:26.248868   71582 streamwatcher.go:109] Unable to decode an event from the watch stream: got short buffer with n=0, base=14551, cap=40960
W0514 21:35:26.248897   71582 reflector.go:323] github.com/openshift/origin/pkg/authorization/generated/informers/internalversion/factory.go:45: watch of *api.ClusterRoleBinding ended with: very short watch
E0514 21:35:26.741940   71582 streamwatcher.go:109] Unable to decode an event from the watch stream: got short buffer with n=0, base=491, cap=2560
W0514 21:35:26.741965   71582 reflector.go:323] github.com/openshift/origin/pkg/authorization/generated/informers/internalversion/factory.go:45: watch of *api.Role ended with: very short watch
E0514 21:35:27.016531   71582 streamwatcher.go:109] Unable to decode an event from the watch stream: got short buffer with n=0, base=38185, cap=163840
W0514 21:35:27.016555   71582 reflector.go:323] github.com/openshift/origin/pkg/authorization/generated/informers/internalversion/factory.go:45: watch of *api.ClusterRole ended with: very short watch
E0514 21:35:27.253762   71582 streamwatcher.go:109] Unable to decode an event from the watch stream: got short buffer with n=0, base=14551, cap=40960
E0514 21:35:27.253817   71582 streamwatcher.go:109] Unable to decode an event from the watch stream: got short buffer with n=0, base=13830, cap=40960
```

**Comment by smarterclayton** [comment]
Or the watch is failing right away because of bad logic somewhere?  Maybe the client gets 404 on watch (not supported, returns list) and then chokes and dies unexpectedly?

**Comment by deads2k** [comment]
> Looks like this is looping trying to WATCH things that don't support watches:

Well, that sucks.

**Comment by deads2k** [comment]
`streamwatcher.go:109] Unable to decode an event from the watch stream: no kind "ClusterRoleList" is registered for version "v1"`

I wonder how long we've been busted

**Comment by deads2k** [comment]
Also, I hate virtual resources

**Comment by liggitt** [comment]
> I wonder how long we've been busted

when Watch collapsed into calling List with `watch:true` in the ListOptions

**Pull request merged**